quickwit/quickwit-proto/src/indexing/mod.rs

```diff
@@ -209,13 +210,146 @@ impl TryFrom<&str> for IndexingTask {
+    pub const fn from_cpu_millis(cpu_millis: u32) -> CpuCapacity {
+        CpuCapacity(cpu_millis)
+    }
+
+    #[inline(always)]
+    pub fn cpu_millis(self) -> u32 {
+        self.0
+    }
+
+    #[inline(always)]
+    pub fn zero() -> CpuCapacity {
+        CpuCapacity::from_cpu_millis(0u32)
+    }
+
+    #[inline(always)]
+    pub fn one_cpu_thread() -> CpuCapacity {
+        CpuCapacity::from_cpu_millis(1_000u32)
+    }
+}
+
+impl Sub<CpuCapacity> for CpuCapacity {
+    type Output = CpuCapacity;
+
+    #[inline(always)]
+    fn sub(self, rhs: CpuCapacity) -> Self::Output {
+        CpuCapacity::from_cpu_millis(self.0 - rhs.0)
+    }
+}
+
+impl Add<CpuCapacity> for CpuCapacity {
+    type Output = CpuCapacity;
+
+    #[inline(always)]
+    fn add(self, rhs: CpuCapacity) -> Self::Output {
+        CpuCapacity::from_cpu_millis(self.0 + rhs.0)
+    }
+}
+
+impl Mul<u32> for CpuCapacity {
+    type Output = CpuCapacity;
+
+    #[inline(always)]
+    fn mul(self, rhs: u32) -> CpuCapacity {
+        CpuCapacity::from_cpu_millis(self.0 * rhs)
+    }
+}
+
+impl Mul<f32> for CpuCapacity {
+    type Output = CpuCapacity;
+
+    #[inline(always)]
+    fn mul(self, scale: f32) -> CpuCapacity {
+        CpuCapacity::from_cpu_millis((self.0 as f32 * scale) as u32)
+    }
+}
+
+impl Display for CpuCapacity {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        write!(f, "{}m", self.0)
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+#[serde(untagged)]
+enum CpuCapacityForSerialization {
+    Float(f32),
+    MilliCpuWithUnit(String),
+}
+
+impl TryFrom<CpuCapacityForSerialization> for CpuCapacity {
+    type Error = String;
+
+    fn try_from(
+        cpu_capacity_for_serialization: CpuCapacityForSerialization,
+    ) -> Result<CpuCapacity, Self::Error> {
+        match cpu_capacity_for_serialization {
+            CpuCapacityForSerialization::Float(cpu_capacity) => {
+                Ok(CpuCapacity((cpu_capacity * 1000.0f32) as u32))
+            }
+            CpuCapacityForSerialization::MilliCpuWithUnit(cpu_capacity_str) => {
+                Self::from_str(&cpu_capacity_str)
+            }
+        }
+    }
+}
+
+impl FromStr for CpuCapacity {
+    type Err = String;
+
+    fn from_str(cpu_capacity_str: &str) -> Result<Self, Self::Err> {
+        let Some(milli_cpus_without_unit_str) = cpu_capacity_str.strip_suffix('m') else {
+            return Err(format!(
+                "invalid cpu capacity: `{cpu_capacity_str}`. String format expects a trailing 'm'."
+            ));
+        };
+        let milli_cpus: u32 = milli_cpus_without_unit_str
+            .parse::<u32>()
+            .map_err(|_err| format!("invalid cpu capacity: `{cpu_capacity_str}`."))?;
+        Ok(CpuCapacity(milli_cpus))
+    }
+}
+
+impl From<CpuCapacity> for CpuCapacityForSerialization {
+    fn from(cpu_capacity: CpuCapacity) -> CpuCapacityForSerialization {
+        CpuCapacityForSerialization::MilliCpuWithUnit(format!("{}m", cpu_capacity.0))
```

Comment:
I know it's tested in `quickwit-config`, but this could be tested in this file directly.

Reply:
adding tests